File: text_extractor.ts
// Airbnb Text Extractor em Deno
// Substituição para o text_extractor.py

import { parse } from "https://deno.land/std@0.192.0/flags/mod.ts";
import * as path from "https://deno.land/std@0.192.0/path/mod.ts";
import { ensureDir } from "https://deno.land/std@0.192.0/fs/ensure_dir.ts";
import { SupabaseUpdater } from "./supabase_updater.ts";

interface ExtractionResult {
  roomId: string;
  timestamp: string;
  content: string;
  metadata: {
    title?: string;
    price?: string;
    rating?: string;
    reviews?: string;
    location?: string;
    features?: {
      rooms?: string;
      bathrooms?: string;
      guests?: string;
    };
    originalFilename?: string;
  };
}

/**
 * Classe para extrair e organizar o conteúdo textual de páginas HTML do Airbnb.
 */
class AirbnbTextExtractor {
  /**
   * Inicializa o extrator.
   */
  constructor() {
    console.log("Inicializando extrator de texto do Airbnb em Deno");
  }

  /**
   * Extrai o conteúdo textual de um arquivo HTML do Airbnb.
   * @param htmlFilePath Caminho para o arquivo HTML
   * @returns Objeto com o conteúdo extraído e metadados
   */
  async extractFromFile(htmlFilePath: string): Promise<ExtractionResult | null> {
    try {
      console.log(`Processando: ${htmlFilePath}`);
      
      // Ler o arquivo HTML
      const htmlContent = await Deno.readTextFile(htmlFilePath);
      if (!htmlContent || htmlContent.trim().length === 0) {
        console.error(`Arquivo vazio ou inválido: ${htmlFilePath}`);
        return null;
      }
      
      // Obter o nome do arquivo
      const filename = path.basename(htmlFilePath);
      
      // Extrair ID do quarto e timestamp do nome do arquivo
      const { roomId, timestamp } = this._extractInfoFromFilename(filename);
      if (!roomId) {
        console.error(`Não foi possível extrair o ID do quarto do nome do arquivo: ${filename}`);
        return null;
      }
      
      console.log(`Extraindo dados para o quarto ID: ${roomId}, timestamp: ${timestamp}`);

      // Extrair o conteúdo relevante do HTML usando expressões regulares básicas
      // Nota: Isto é uma simplificação, idealmente usaríamos um parser HTML completo
      const title = this._extractTitle(htmlContent);
      const price = this._extractPrice(htmlContent);
      const { rating, reviews } = this._extractReviews(htmlContent);
      const location = this._extractLocation(htmlContent);
      const features = this._extractFeatures(htmlContent);
      
      // Criar o objeto de resultado
      const result: ExtractionResult = {
        roomId,
        timestamp,
        content: this._extractMainContent(htmlContent),
        metadata: {
          title,
          price,
          rating,
          reviews,
          location,
          features,
          originalFilename: filename
        }
      };
      
      console.log(`✅ Extração concluída para o quarto ${roomId}`);
      if (title) console.log(`   Título: ${title}`);
      if (price) console.log(`   Preço: ${price}`);
      
      return result;
    } catch (error) {
      console.error(`Erro ao processar o arquivo ${htmlFilePath}: ${error.message}`);
      return null;
    }
  }

  /**
   * Processa todos os arquivos HTML em um diretório.
   * @param inputDir Diretório com arquivos HTML
   * @param outputDir Diretório para salvar os arquivos de saída
   * @param format Formato de saída ('json', 'text', 'both')
   * @param updateSupabase Se deve atualizar os títulos no Supabase
   * @returns Número de arquivos processados com sucesso
   */
  async processDirectory(
    inputDir: string, 
    outputDir: string, 
    format: string = "both",
    updateSupabase: boolean = false
  ): Promise<number> {
    try {
      // Garantir que o diretório de saída existe
      await ensureDir(outputDir);
      
      // Listar todos os arquivos HTML no diretório de entrada
      const files = [];
      for await (const entry of Deno.readDir(inputDir)) {
        if (entry.isFile && entry.name.endsWith(".html")) {
          files.push(entry.name);
        }
      }
      
      if (files.length === 0) {
        console.log(`Nenhum arquivo HTML encontrado em ${inputDir}`);
        return 0;
      }
      
      console.log(`Encontrados ${files.length} arquivos HTML para processar`);
      
      // Inicializar o atualizador Supabase se necessário
      let updater: SupabaseUpdater | null = null;
      if (updateSupabase) {
        try {
          updater = new SupabaseUpdater();
          console.log("Cliente Supabase inicializado para atualizações de títulos");
        } catch (error) {
          console.error(`Erro ao inicializar cliente Supabase: ${error.message}`);
          console.log("Continuando sem atualizar o Supabase");
        }
      }
      
      // Processar cada arquivo
      let successCount = 0;
      
      for (const filename of files) {
        const filePath = path.join(inputDir, filename);
        const result = await this.extractFromFile(filePath);
        
        if (result) {
          // Salvar o resultado no formato especificado
          if (format === "json" || format === "both") {
            const jsonPath = path.join(outputDir, `airbnb_${result.roomId}_${result.timestamp}.json`);
            await Deno.writeTextFile(jsonPath, JSON.stringify(result, null, 2));
            console.log(`JSON salvo: ${jsonPath}`);
          }
          
          if (format === "text" || format === "both") {
            const textPath = path.join(outputDir, `airbnb_${result.roomId}_${result.timestamp}.txt`);
            await Deno.writeTextFile(textPath, result.content);
            console.log(`Texto salvo: ${textPath}`);
          }
          
          // Atualizar o Supabase se necessário
          if (updateSupabase && updater && result.metadata.title) {
            try {
              const updateResult = await updater.updateRoomLabel(result.roomId, result.metadata.title);
              console.log(`Supabase ${updateResult.success ? 'atualizado' : 'falhou'} para o quarto ${result.roomId}`);
            } catch (error) {
              console.error(`Erro ao atualizar Supabase para quarto ${result.roomId}: ${error.message}`);
            }
          }
          
          successCount++;
        }
      }
      
      console.log(`✅ Processamento concluído. ${successCount} de ${files.length} arquivos processados com sucesso.`);
      return successCount;
    } catch (error) {
      console.error(`Erro ao processar diretório: ${error.message}`);
      return 0;
    }
  }

  /**
   * Extrai o conteúdo principal do HTML
   * @param html Conteúdo HTML
   * @returns Texto extraído
   */
  private _extractMainContent(html: string): string {
    // Remover tags de script e style
    let content = html.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, ' ');
    content = content.replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, ' ');
    
    // Remover todas as tags HTML mantendo o texto
    content = content.replace(/<[^>]+>/g, ' ');
    
    // Normalizar espaços em branco
    content = content.replace(/\s+/g, ' ').trim();
    
    // Nota: isso é uma simplificação básica. Uma versão mais avançada
    // faria uma análise semântica para extrair apenas as partes importantes.
    
    return content;
  }

  /**
   * Extrai informações do nome do arquivo
   * @param filename Nome do arquivo
   * @returns Objeto com roomId e timestamp
   */
  private _extractInfoFromFilename(filename: string): { roomId: string; timestamp: string } {
    // Formato esperado: airbnb_ROOMID_TIMESTAMP.html
    const match = filename.match(/airbnb_(\d+)_(.+)\.html$/);
    
    if (match && match.length >= 3) {
      return {
        roomId: match[1],
        timestamp: match[2]
      };
    }
    
    return { roomId: "", timestamp: "" };
  }

  /**
   * Extrai o título do anúncio
   * @param html Conteúdo HTML
   * @returns Título extraído ou undefined
   */
  private _extractTitle(html: string): string | undefined {
    // Primeira tentativa: meta tag de título
    const metaTitleMatch = html.match(/<meta\s+property="og:title"\s+content="([^"]+)"/i);
    if (metaTitleMatch && metaTitleMatch[1]) {
      return metaTitleMatch[1].trim();
    }
    
    // Segunda tentativa: tag h1
    const h1Match = html.match(/<h1[^>]*>([^<]+)<\/h1>/i);
    if (h1Match && h1Match[1]) {
      return h1Match[1].trim();
    }
    
    // Terceira tentativa: padrão de dados JSON embutido
    const jsonMatch = html.match(/"title":"([^"]+)"/);
    if (jsonMatch && jsonMatch[1]) {
      return jsonMatch[1].trim();
    }
    
    return undefined;
  }

  /**
   * Extrai o preço do anúncio
   * @param html Conteúdo HTML
   * @returns Preço extraído ou undefined
   */
  private _extractPrice(html: string): string | undefined {
    // Buscar por padrões comuns de preço no HTML
    const priceMatch = html.match(/"priceString":"([^"]+)"/);
    if (priceMatch && priceMatch[1]) {
      return priceMatch[1].trim();
    }
    
    // Alternativas
    const altPriceMatch = html.match(/R\$\s*[\d.]+/);
    if (altPriceMatch) {
      return altPriceMatch[0].trim();
    }
    
    return undefined;
  }

  /**
   * Extrai a avaliação e quantidade de reviews
   * @param html Conteúdo HTML
   * @returns Objeto com rating e reviews
   */
  private _extractReviews(html: string): { rating?: string; reviews?: string } {
    const result: { rating?: string; reviews?: string } = {};
    
    // Buscar avaliação
    const ratingMatch = html.match(/"starRating":([0-9.]+)/);
    if (ratingMatch && ratingMatch[1]) {
      result.rating = ratingMatch[1].trim();
    }
    
    // Buscar número de reviews
    const reviewsMatch = html.match(/"reviewCount":([0-9]+)/);
    if (reviewsMatch && reviewsMatch[1]) {
      result.reviews = reviewsMatch[1].trim();
    }
    
    return result;
  }

  /**
   * Extrai a localização do anúncio
   * @param html Conteúdo HTML
   * @returns Localização extraída ou undefined
   */
  private _extractLocation(html: string): string | undefined {
    // Buscar localização na meta tag
    const locationMatch = html.match(/<meta\s+property="og:location"\s+content="([^"]+)"/i);
    if (locationMatch && locationMatch[1]) {
      return locationMatch[1].trim();
    }
    
    // Alternativa: buscar no JSON embutido
    const jsonLocationMatch = html.match(/"location":\s*{[^}]*"city":"([^"]+)"/);
    if (jsonLocationMatch && jsonLocationMatch[1]) {
      return jsonLocationMatch[1].trim();
    }
    
    return undefined;
  }

  /**
   * Extrai características do imóvel (quartos, banheiros, etc.)
   * @param html Conteúdo HTML
   * @returns Objeto com características
   */
  private _extractFeatures(html: string): { rooms?: string; bathrooms?: string; guests?: string } {
    const features: { rooms?: string; bathrooms?: string; guests?: string } = {};
    
    // Quartos
    const roomsMatch = html.match(/"bedrooms":([0-9]+)/);
    if (roomsMatch && roomsMatch[1]) {
      features.rooms = roomsMatch[1].trim();
    }
    
    // Banheiros
    const bathroomsMatch = html.match(/"bathrooms":([0-9.]+)/);
    if (bathroomsMatch && bathroomsMatch[1]) {
      features.bathrooms = bathroomsMatch[1].trim();
    }
    
    // Hóspedes
    const guestsMatch = html.match(/"personCapacity":([0-9]+)/);
    if (guestsMatch && guestsMatch[1]) {
      features.guests = guestsMatch[1].trim();
    }
    
    return features;
  }
}

/**
 * Função principal do programa
 */
async function main(): Promise<void> {
  // Analisar argumentos da linha de comando
  const args = parse(Deno.args, {
    string: ["input-dir", "output-dir", "format", "input"],
    boolean: ["help", "update-supabase"],
    default: {
      "input-dir": "./output",
      "output-dir": "./extracted",
      "format": "both",
      "update-supabase": false
    }
  });

  // Mostrar ajuda se solicitado
  if (args.help) {
    console.log("Uso: deno run --allow-read --allow-write --allow-env text_extractor.ts [OPÇÕES]");
    console.log("");
    console.log("Opções:");
    console.log("  --input-dir=DIR       Diretório com arquivos HTML (padrão: ./output)");
    console.log("  --output-dir=DIR      Diretório para salvar os resultados (padrão: ./extracted)");
    console.log("  --format=FORMAT       Formato de saída: json, text, both (padrão: both)");
    console.log("  --update-supabase     Atualiza títulos extraídos no Supabase");
    console.log("  --help                Mostra esta ajuda");
    Deno.exit(0);
  }

  // Verificar entrada
  const inputDir = args["input-dir"];
  const outputDir = args["output-dir"];
  const format = args["format"];
  const updateSupabase = args["update-supabase"];

  try {
    console.log(`Iniciando extração de texto do Airbnb (Deno)`);
    console.log(`Diretório de entrada: ${inputDir}`);
    console.log(`Diretório de saída: ${outputDir}`);
    console.log(`Formato: ${format}`);
    console.log(`Atualizar Supabase: ${updateSupabase ? "Sim" : "Não"}`);
    console.log("");

    // Verificar se o diretório de entrada existe
    try {
      const stat = await Deno.stat(inputDir);
      if (!stat.isDirectory) {
        console.error(`Erro: ${inputDir} não é um diretório.`);
        Deno.exit(1);
      }
    } catch (error) {
      console.error(`Erro: Diretório de entrada '${inputDir}' não existe.`);
      Deno.exit(1);
    }

    // Inicializar o extrator e processar o diretório
    const extractor = new AirbnbTextExtractor();
    const processedCount = await extractor.processDirectory(
      inputDir,
      outputDir,
      format,
      updateSupabase
    );

    if (processedCount > 0) {
      console.log(`✅ Extração concluída! ${processedCount} arquivos processados.`);
      Deno.exit(0);
    } else {
      console.error("❌ Nenhum arquivo foi processado com sucesso.");
      Deno.exit(1);
    }
  } catch (error) {
    console.error(`Erro durante a execução: ${error.message}`);
    Deno.exit(1);
  }
}

// Executar a função principal se este arquivo for chamado diretamente
if (import.meta.main) {
  main();
}

File: text_extractor.py
#!/usr/bin/env python3
"""
Airbnb Text Extractor

Este script extrai o conteúdo de texto relevante de arquivos HTML de anúncios do Airbnb
usando a biblioteca trafilatura.

Uso:
  python text_extractor.py --input arquivo.html
  python text_extractor.py --input-dir diretorio_de_entrada --output-dir diretorio_de_saida
"""

import os
import sys
import argparse
import glob
import json
from datetime import datetime
from pathlib import Path
import trafilatura
from bs4 import BeautifulSoup

class AirbnbTextExtractor:
    """
    Classe para extrair e organizar o conteúdo textual de páginas HTML do Airbnb.
    """
    
    def __init__(self):
        """Inicializa o extrator."""
        pass
    
    def extract_from_file(self, html_file_path):
        """
        Extrai o conteúdo textual de um arquivo HTML do Airbnb.
        
        Args:
            html_file_path: Caminho para o arquivo HTML
            
        Returns:
            dict: Dicionário com o conteúdo extraído e metadados
        """
        try:
            # Abrir e ler o arquivo HTML
            with open(html_file_path, 'r', encoding='utf-8') as file:
                html_content = file.read()
            
            # Extrair o conteúdo usando trafilatura com configuração otimizada
            text_content = trafilatura.extract(
                html_content,
                include_comments=False,
                include_tables=True,
                no_fallback=False,  # Usar fallback extraction se necessário
                output_format="txt"  # Corrigido: deve ser 'txt' não 'text'
            )
            
            # Se trafilatura falhar em extrair conteúdo significativo, tentar usar BeautifulSoup para uma extração mais simples
            if not text_content or len(text_content.strip()) < 100:
                print(f"Aviso: Trafilatura extraiu pouco conteúdo, usando fallback com BeautifulSoup")
                soup_fallback = BeautifulSoup(html_content, 'html.parser')
                
                # Verificar se é uma página de erro do Airbnb (JavaScript desabilitado)
                js_error_message = soup_fallback.find(string=lambda text: 'não funcionam corretamente sem a habilitação do JavaScript' in str(text) if text else False)
                if js_error_message:
                    print("Detectada mensagem de erro de JavaScript do Airbnb. O site pode estar bloqueando nosso scraper.")
                
                # Remover scripts, estilos e outros elementos não relevantes
                for element in soup_fallback(['script', 'style', 'head', 'meta', 'link']):
                    element.decompose()
                
                # Tentar encontrar o conteúdo principal
                main_content = soup_fallback.find('main')
                if main_content:
                    text_content = main_content.get_text(separator='\n', strip=True)
                else:
                    # Extrair texto de tudo o que restou se não encontrar o conteúdo principal
                    text_content = soup_fallback.get_text(separator='\n', strip=True)
                
                # Limpar texto extraído
                text_content = '\n'.join(line.strip() for line in text_content.splitlines() if line.strip())
            
            # Extrair informações estruturadas com BeautifulSoup
            soup = BeautifulSoup(html_content, 'html.parser')
            
            # Tentar extrair o título da propriedade
            title = self._extract_title(soup)
            
            # Tentar extrair o preço
            price = self._extract_price(soup)
            
            # Tentar extrair avaliações
            rating, review_count = self._extract_reviews(soup)
            
            # Tentar extrair localização
            location = self._extract_location(soup)
            
            # Tentar extrair características (quartos, banheiros, etc.)
            features = self._extract_features(soup)
            
            # Extrair data do arquivo (do nome do arquivo ou do sistema)
            timestamp = self._extract_timestamp_from_filename(html_file_path)
            
            # Organizar os dados
            result = {
                "metadata": {
                    "source_file": os.path.basename(html_file_path),
                    "extraction_date": datetime.now().isoformat(),
                    "timestamp": timestamp,
                },
                "listing": {
                    "title": title,
                    "price": price,
                    "rating": rating,
                    "review_count": review_count,
                    "location": location,
                    "features": features
                },
                "content": text_content
            }
            
            return result
            
        except Exception as e:
            print(f"Erro ao processar o arquivo {html_file_path}: {str(e)}")
            return None
    
    def process_directory(self, input_dir, output_dir):
        """
        Processa todos os arquivos HTML em um diretório.
        
        Args:
            input_dir: Diretório com arquivos HTML
            output_dir: Diretório para salvar os arquivos de saída
        
        Returns:
            int: Número de arquivos processados com sucesso
        """
        # Criar diretório de saída se não existir
        os.makedirs(output_dir, exist_ok=True)
        
        # Encontrar todos os arquivos HTML no diretório
        html_files = glob.glob(os.path.join(input_dir, "*.html"))
        
        if not html_files:
            print(f"Nenhum arquivo HTML encontrado em {input_dir}")
            return 0
        
        success_count = 0
        
        for html_file in html_files:
            # Extrair ID do quarto do nome do arquivo
            room_id = self._extract_room_id_from_filename(html_file)
            
            # Nome base do arquivo (sem extensão)
            base_name = os.path.splitext(os.path.basename(html_file))[0]
            
            # Processar o arquivo
            result = self.extract_from_file(html_file)
            
            if result:
                # Salvar como JSON
                json_output_path = os.path.join(output_dir, f"{base_name}.json")
                with open(json_output_path, 'w', encoding='utf-8') as json_file:
                    json.dump(result, json_file, ensure_ascii=False, indent=2)
                
                # Salvar conteúdo de texto em arquivo separado
                text_output_path = os.path.join(output_dir, f"{base_name}.txt")
                with open(text_output_path, 'w', encoding='utf-8') as text_file:
                    text_file.write(result["content"] if result["content"] else "")
                
                print(f"Processado: {html_file} -> {json_output_path}")
                success_count += 1
        
        return success_count

    def _extract_title(self, soup):
        """Extrai o título do anúncio."""
        # Tentar diferentes seletores para o título
        title_selectors = [
            'h1[data-testid="listing-title"]',
            'h1._fecoyn4',
            'h1.atm_7l_1kw7nm4',
            'h2.atm_7l_1kw7nm4',
            'h1',  # Fallback genérico: primeiro h1 da página
            'title'  # Fallback último recurso: usar o título da página
        ]
        
        for selector in title_selectors:
            title_elem = soup.select_one(selector)
            if title_elem:
                title_text = title_elem.get_text(strip=True)
                # Se for o título da página, tentar limpar um pouco
                if selector == 'title' and ' - Airbnb' in title_text:
                    title_text = title_text.split(' - Airbnb')[0].strip()
                return title_text
        
        return None
    
    def _extract_price(self, soup):
        """Extrai o preço do anúncio."""
        # Tentar diferentes seletores para o preço
        price_selectors = [
            'span[data-testid="listing-price"]',
            'span._tyxjp1',
            'span.atm_c8_1n3c8jb'
        ]
        
        for selector in price_selectors:
            price_elem = soup.select_one(selector)
            if price_elem:
                return price_elem.get_text(strip=True)
        
        return None
    
    def _extract_reviews(self, soup):
        """Extrai a avaliação e quantidade de reviews."""
        rating = None
        review_count = None
        
        # Tentar extrair a avaliação
        rating_selectors = [
            'span[data-testid="rating-value"]',
            'span._17p6nbba',
            'span.atm_7l_1kw7nm4 span'
        ]
        
        for selector in rating_selectors:
            rating_elem = soup.select_one(selector)
            if rating_elem:
                try:
                    rating = float(rating_elem.get_text(strip=True).replace(',', '.'))
                    break
                except ValueError:
                    pass
        
        # Tentar extrair o número de avaliações
        review_count_selectors = [
            'span[data-testid="reviews-count"]',
            'span._s65ijh7',
            'button._1qf7wt0z span'
        ]
        
        for selector in review_count_selectors:
            count_elem = soup.select_one(selector)
            if count_elem:
                text = count_elem.get_text(strip=True)
                # Extrair apenas os números
                import re
                numbers = re.findall(r'\d+', text)
                if numbers:
                    try:
                        review_count = int(''.join(numbers))
                        break
                    except ValueError:
                        pass
        
        return rating, review_count
    
    def _extract_location(self, soup):
        """Extrai a localização do anúncio."""
        # Tentar diferentes seletores para a localização
        location_selectors = [
            'span[data-testid="listing-location"]',
            'div._9bezani',
            'div.atm_c8_exq9by5'
        ]
        
        for selector in location_selectors:
            location_elem = soup.select_one(selector)
            if location_elem:
                return location_elem.get_text(strip=True)
        
        return None
    
    def _extract_features(self, soup):
        """Extrai características do imóvel (quartos, banheiros, etc.)."""
        features = {}
        
        # Tentar extrair as características principais
        feature_selectors = [
            'div[data-testid="listing-features"]',
            'div._1dotkqq',
            'div.atm_cs_1pkprkl'
        ]
        
        for selector in feature_selectors:
            feature_container = soup.select_one(selector)
            if feature_container:
                feature_text = feature_container.get_text(strip=True)
                
                # Tenta identificar padrões como "2 quartos", "1 banheiro", etc.
                import re
                rooms = re.search(r'(\d+)\s*quarto', feature_text)
                baths = re.search(r'(\d+)\s*banheiro', feature_text)
                beds = re.search(r'(\d+)\s*cama', feature_text)
                
                if rooms:
                    features['rooms'] = int(rooms.group(1))
                if baths:
                    features['bathrooms'] = int(baths.group(1))
                if beds:
                    features['beds'] = int(beds.group(1))
                
                break
        
        return features
    
    def _extract_timestamp_from_filename(self, filename):
        """Extrai o timestamp do nome do arquivo."""
        # Padrão esperado: airbnb_ROOMID_YYYY-MM-DDTHH-MM-SS.html
        try:
            import re
            match = re.search(r'(\d{4}-\d{2}-\d{2}T\d{2}-\d{2}-\d{2})', filename)
            if match:
                timestamp_str = match.group(1)
                # Converter para formato ISO 8601
                timestamp_str = timestamp_str.replace('-', ':', 2)
                return timestamp_str
        except Exception:
            pass
        
        # Fallback: usar a data de modificação do arquivo
        return datetime.fromtimestamp(os.path.getmtime(filename)).isoformat()
    
    def _extract_room_id_from_filename(self, filename):
        """Extrai o ID do quarto do nome do arquivo."""
        # Padrão esperado: airbnb_ROOMID_YYYY-MM-DDTHH-MM-SS.html
        try:
            import re
            match = re.search(r'airbnb_(\d+)_', os.path.basename(filename))
            if match:
                return match.group(1)
        except Exception:
            pass
        
        return None

def main():
    """Função principal do programa."""
    parser = argparse.ArgumentParser(description='Extrai conteúdo de texto de arquivos HTML do Airbnb.')
    
    input_group = parser.add_mutually_exclusive_group(required=True)
    input_group.add_argument('--input', help='Caminho para um arquivo HTML específico')
    input_group.add_argument('--input-dir', help='Diretório contendo arquivos HTML para processar')
    
    parser.add_argument('--output-dir', default='./extracted', help='Diretório para salvar os arquivos de saída')
    parser.add_argument('--format', choices=['json', 'text', 'both'], default='both', 
                       help='Formato de saída (json, text ou both)')
    
    args = parser.parse_args()
    
    extractor = AirbnbTextExtractor()
    
    if args.input:
        # Processar um único arquivo
        result = extractor.extract_from_file(args.input)
        
        if not result:
            print(f"Falha ao processar o arquivo {args.input}")
            return 1
        
        # Criar diretório de saída se não existir
        os.makedirs(args.output_dir, exist_ok=True)
        
        # Nome base do arquivo (sem extensão)
        base_name = os.path.splitext(os.path.basename(args.input))[0]
        
        # Salvar resultado
        if args.format in ['json', 'both']:
            json_output_path = os.path.join(args.output_dir, f"{base_name}.json")
            with open(json_output_path, 'w', encoding='utf-8') as json_file:
                json.dump(result, json_file, ensure_ascii=False, indent=2)
            print(f"Arquivo JSON salvo em: {json_output_path}")
        
        if args.format in ['text', 'both']:
            text_output_path = os.path.join(args.output_dir, f"{base_name}.txt")
            with open(text_output_path, 'w', encoding='utf-8') as text_file:
                text_file.write(result["content"] if result["content"] else "")
            print(f"Arquivo de texto salvo em: {text_output_path}")
    
    else:
        # Processar um diretório
        count = extractor.process_directory(args.input_dir, args.output_dir)
        print(f"Processados {count} arquivos HTML de {args.input_dir} para {args.output_dir}")
    
    return 0

if __name__ == "__main__":
    sys.exit(main())